From 414e5fdcb81b8507754e52e0a614713dbef9ef4f Mon Sep 17 00:00:00 2001
From: haiqiang yao <haiqiangx.yao@intel.com>
Date: Wed, 30 Dec 2015 15:16:11 +0800
Subject: [PATCH 03/13] atomisp2: Add some new parameters for isp2.7

this patch will add some new parameters for isp2.7.

Change-Id: Ib5e4164d3493c8e734cb3e23fa111eb4005e5d2b
Original-Tracked_On: #H1504135600
Signed-off-by: haiqiang yao <haiqiangx.yao@intel.com>
Reviewed-on: https://icggerrit.ir.intel.com/95171
Approver: Bin B Han <bin.b.han@intel.com>
Reviewed-by: Bin B Han <bin.b.han@intel.com>
Validation-Android: Wei Tao <wei.tao@intel.com>
Build: Dessert M BXT RVP (SC)
Integrator: WeiX Meng <weix.meng@intel.com>
Reviewed-by: WeiX Meng <weix.meng@intel.com>
Tracked-On: https://jira01.devtools.intel.com/browse/OAM-17482
Signed-off-by: Tao Wei <wei.tao@intel.com>
Reviewed-on: https://android.intel.com:443/464446
---
 .../pci/atomisp2/atomisp_driver/atomisp_cmd.c      | 103 ++++++++++++++++++++-
 .../atomisp2/atomisp_driver/atomisp_compat_css20.c |  41 ++++++++
 .../atomisp2/atomisp_driver/atomisp_compat_css20.h |  31 ++++++-
 .../atomisp_driver/atomisp_compat_ioctl32.h        |   9 ++
 .../pci/atomisp2/atomisp_driver/atomisp_subdev.h   |  11 +++
 5 files changed, 193 insertions(+), 2 deletions(-)

diff --git a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_cmd.c b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_cmd.c
index a4b545f..acddfa0 100644
--- a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_cmd.c
+++ b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_cmd.c
@@ -3236,7 +3236,7 @@ void atomisp_apply_css_parameters(
 		atomisp_css_set_dvs_6axis(asd, css_param->dvs_6axis);
 
 	/* Workaround for BTNS, apply default OB2 setting for IMX227.
-	   before AIQ for ISP2.7 is available,	we have to keep such
+	   before AIQ for ISP2.7 is available,  we have to keep such
 	   workaround for a reasonable image color
 	 */
 #ifdef CONFIG_EXTERNAL_BTNS_CAMERA
@@ -3246,6 +3246,29 @@ void atomisp_apply_css_parameters(
 		atomisp_css_set_ob2_config(asd, &imx227_ob2_config);
 	}
 #endif
+
+	/* Add some parameters for isp2.7 */
+	if (css_param->update_flag.dpc2_config)
+		atomisp_css_set_dpc2_config(asd, &css_param->dpc2_config);
+
+	if (css_param->update_flag.eed1_8_config)
+		atomisp_css_set_eed1_8_config(asd, &css_param->eed1_8_config);
+
+	if (css_param->update_flag.ctc2_config)
+		atomisp_css_set_ctc2_config(asd, &css_param->ctc2_config);
+
+	if (css_param->update_flag.iefd2_6_config)
+		atomisp_css_set_iefd2_6_config(asd, &css_param->iefd2_6_config);
+
+	if (css_param->update_flag.macc1_5_config)
+		atomisp_css_set_macc1_5_config(asd, &css_param->macc1_5_config);
+
+	if (css_param->update_flag.macc1_5_table)
+		atomisp_css_set_macc1_5_table(asd, &css_param->macc1_5_table);
+
+	if (css_param->update_flag.xnr3_0_11_config)
+		atomisp_css_set_xnr3_0_11_config(asd, &css_param->xnr3_0_11_config);
+
 	atomisp_css_set_isp_config_id(asd, css_param->isp_config_id);
 	/*
 	 * These configurations are on used by ISP1.x, not for ISP2.x,
@@ -3547,6 +3570,84 @@ int atomisp_cp_general_isp_parameters(struct atomisp_sub_device *asd,
 			(struct atomisp_anr_thres *) &css_param->anr_thres;
 	}
 
+	if (arg->dpc2_config && (from_user || !cur_config->dpc2_config)) {
+		if (copy_from_compatible(&css_param->dpc2_config,
+				arg->dpc2_config,
+				sizeof(struct atomisp_css_dpc2_config),
+				from_user))
+			return -EFAULT;
+		css_param->update_flag.dpc2_config =
+			(struct atomisp_dpc2_config *) &css_param->dpc2_config;
+	}
+
+	if (arg->eed1_8_config && (from_user || !cur_config->eed1_8_config)) {
+		if (copy_from_compatible(&css_param->eed1_8_config,
+				arg->eed1_8_config,
+				sizeof(struct atomisp_css_eed1_8_config),
+				from_user))
+			return -EFAULT;
+		css_param->update_flag.eed1_8_config =
+			(struct atomisp_eed1_8_config *) &css_param->eed1_8_config;
+	}
+
+	if (arg->ob2_config && (from_user || !cur_config->ob2_config)) {
+		if (copy_from_compatible(&css_param->ob2_config, arg->ob2_config,
+				sizeof(struct atomisp_css_ob2_config),
+				from_user))
+			return -EFAULT;
+		css_param->update_flag.ob2_config =
+			(struct atomisp_ob2_config *) &css_param->ob2_config;
+	}
+
+	if (arg->ctc2_config && (from_user || !cur_config->ctc2_config)) {
+		if (copy_from_compatible(&css_param->ctc2_config, arg->ctc2_config,
+				sizeof(struct atomisp_css_ctc2_config),
+				from_user))
+			return -EFAULT;
+		css_param->update_flag.ctc2_config =
+			(struct atomisp_ctc2_config *) &css_param->ctc2_config;
+	}
+
+	if (arg->iefd2_6_config && (from_user || !cur_config->iefd2_6_config)) {
+		if (copy_from_compatible(&css_param->iefd2_6_config,
+				arg->iefd2_6_config,
+				sizeof(struct atomisp_css_iefd2_6_config),
+				from_user))
+			return -EFAULT;
+		css_param->update_flag.iefd2_6_config =
+			(struct atomisp_iefd2_6_config *) &css_param->iefd2_6_config;
+	}
+
+	if (arg->macc1_5_config && (from_user || !cur_config->macc1_5_config)) {
+		if (copy_from_compatible(&css_param->macc1_5_config,
+				arg->macc1_5_config,
+				sizeof(struct atomisp_css_macc1_5_config),
+				from_user))
+			return -EFAULT;
+		css_param->update_flag.macc1_5_config =
+			(struct atomisp_macc1_5_config *) &css_param->macc1_5_config;
+	}
+
+	if (arg->macc1_5_table && (from_user || !cur_config->macc1_5_table)) {
+		if (copy_from_compatible(&css_param->macc1_5_table,
+				arg->macc1_5_table,
+				sizeof(struct atomisp_css_macc1_5_table),
+				from_user))
+			return -EFAULT;
+		css_param->update_flag.macc1_5_table =
+			(struct atomisp_macc1_5_table *) &css_param->macc1_5_table;
+	}
+
+	if (arg->xnr3_0_11_config && (from_user || !cur_config->xnr3_0_11_config)) {
+		if (copy_from_compatible(&css_param->xnr3_0_11_config,
+				arg->xnr3_0_11_config,
+				sizeof(struct atomisp_css_xnr3_0_11_config),
+				from_user))
+			return -EFAULT;
+		css_param->update_flag.xnr3_0_11_config =
+			(struct atomisp_xnr3_0_11_config *) &css_param->xnr3_0_11_config;
+	}
+
 	if (from_user)
 		css_param->isp_config_id = arg->isp_config_id;
 	/*
diff --git a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_compat_css20.c b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_compat_css20.c
index 00d5170..e4ec9aa 100644
--- a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_compat_css20.c
+++ b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_compat_css20.c
@@ -3892,12 +3892,53 @@ int atomisp_css_get_ob_config(struct atomisp_sub_device *asd,
 	return 0;
 }
 
+void atomisp_css_set_dpc2_config(struct atomisp_sub_device *asd,
+			struct atomisp_css_dpc2_config *dpc2_config)
+{
+	asd->params.config.dpc2_config = dpc2_config;
+}
+
+void atomisp_css_set_eed1_8_config(struct atomisp_sub_device *asd,
+			struct atomisp_css_eed1_8_config *eed1_8_config)
+{
+	asd->params.config.eed1_8_config = eed1_8_config;
+}
 void atomisp_css_set_ob2_config(struct atomisp_sub_device *asd,
 			struct atomisp_css_ob2_config *ob2_config)
 {
 	asd->params.config.ob2_config = ob2_config;
 }
 
+void atomisp_css_set_ctc2_config(struct atomisp_sub_device *asd,
+			struct atomisp_css_ctc2_config *ctc2_config)
+{
+	asd->params.config.ctc2_config = ctc2_config;
+}
+
+void atomisp_css_set_iefd2_6_config(struct atomisp_sub_device *asd,
+			struct atomisp_css_iefd2_6_config *iefd2_6_config)
+{
+	asd->params.config.iefd2_6_config = iefd2_6_config;
+}
+
+void atomisp_css_set_macc1_5_config(struct atomisp_sub_device *asd,
+			struct atomisp_css_macc1_5_config *macc1_5_config)
+{
+	asd->params.config.macc1_5_config = macc1_5_config;
+}
+
+void atomisp_css_set_macc1_5_table(struct atomisp_sub_device *asd,
+			struct atomisp_css_macc1_5_table *macc1_5_table)
+{
+	asd->params.config.macc1_5_table = macc1_5_table;
+}
+
+void atomisp_css_set_xnr3_0_11_config(struct atomisp_sub_device *asd,
+			struct atomisp_css_xnr3_0_11_config *xnr3_0_11_config)
+{
+	asd->params.config.xnr3_0_11_config = xnr3_0_11_config;
+}
+
 int atomisp_css_get_dp_config(struct atomisp_sub_device *asd,
 			struct atomisp_dp_config *config)
 {
diff --git a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_compat_css20.h b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_compat_css20.h
index 95588c8..826da05 100644
--- a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_compat_css20.h
+++ b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_compat_css20.h
@@ -69,7 +69,6 @@
 #define atomisp_css_anr_config	ia_css_anr_config
 #define atomisp_css_xnr_config	ia_css_xnr_config
 #define atomisp_css_macc_config	ia_css_macc_config
-#define atomisp_css_ob2_config	ia_css_ob2_config
 #define atomisp_css_gamma_table	ia_css_gamma_table
 #define atomisp_css_ctc_table	ia_css_ctc_table
 #define atomisp_css_macc_table	ia_css_macc_table
@@ -85,6 +84,15 @@
 #define atomisp_css_dvs_6axis_config	ia_css_dvs_6axis_config
 #define atomisp_css_fw_info	ia_css_fw_info
 #define atomisp_css_formats_config	ia_css_formats_config
+/*Only add parameters for ISP2.7*/
+#define atomisp_css_ob2_config		ia_css_ob2_config
+#define atomisp_css_dpc2_config	ia_css_dpc2_config
+#define atomisp_css_eed1_8_config	ia_css_eed1_8_config
+#define atomisp_css_ctc2_config		ia_css_ctc2_config
+#define atomisp_css_iefd2_6_config	ia_css_iefd2_6_config
+#define atomisp_css_macc1_5_config	ia_css_macc1_5_config
+#define atomisp_css_macc1_5_table	ia_css_macc1_5_table
+#define atomisp_css_xnr3_0_11_config	ia_css_xnr3_0_11_config
 
 #define CSS_PIPE_ID_PREVIEW	IA_CSS_PIPE_ID_PREVIEW
 #define CSS_PIPE_ID_COPY	IA_CSS_PIPE_ID_COPY
@@ -261,9 +269,30 @@ void atomisp_css_unload_firmware(struct atomisp_device *isp);
 void atomisp_css_set_dvs_6axis(struct atomisp_sub_device *asd,
 			struct atomisp_css_dvs_6axis *dvs_6axis);
 
+void atomisp_css_set_dpc2_config(struct atomisp_sub_device *asd,
+			struct atomisp_css_dpc2_config *dpc2_config);
+
+void atomisp_css_set_eed1_8_config(struct atomisp_sub_device *asd,
+			struct atomisp_css_eed1_8_config *eed1_8_config);
+
 void atomisp_css_set_ob2_config(struct atomisp_sub_device *asd,
 			struct atomisp_css_ob2_config *ob2_config);
 
+void atomisp_css_set_ctc2_config(struct atomisp_sub_device *asd,
+			struct atomisp_css_ctc2_config *ctc2_config);
+
+void atomisp_css_set_iefd2_6_config(struct atomisp_sub_device *asd,
+			struct atomisp_css_iefd2_6_config *iefd2_6_config);
+
+void atomisp_css_set_macc1_5_config(struct atomisp_sub_device *asd,
+			struct atomisp_css_macc1_5_config *macc1_5_config);
+
+void atomisp_css_set_macc1_5_table(struct atomisp_sub_device *asd,
+			struct atomisp_css_macc1_5_table *macc1_5_table);
+
+void atomisp_css_set_xnr3_0_11_config(struct atomisp_sub_device *asd,
+			struct atomisp_css_xnr3_0_11_config *xnr3_0_11_config);
+
 unsigned int atomisp_css_debug_get_dtrace_level(void);
 
 int atomisp_css_debug_dump_isp_binary(void);
diff --git a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_compat_ioctl32.h b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_compat_ioctl32.h
index 750478f..5a556a3 100644
--- a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_compat_ioctl32.h
+++ b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_compat_ioctl32.h
@@ -231,6 +231,15 @@ struct atomisp_parameters32 {
 	compat_uptr_t dvs2_coefs; /* DVS 2.0 coefficients */
 	compat_uptr_t capture_config;
 	compat_uptr_t anr_thres;
+	compat_uptr_t dpc2_config;
+	compat_uptr_t eed1_8_config;
+	compat_uptr_t ob2_config;
+	compat_uptr_t ctc2_config;
+	compat_uptr_t iefd2_6_config;
+	compat_uptr_t macc1_5_config;
+	compat_uptr_t macc1_5_table;
+	compat_uptr_t xnr3_0_11_config;
+
 
 	compat_uptr_t	lin_2500_config;       /* Skylake: Linearization config */
 	compat_uptr_t	obgrid_2500_config;    /* Skylake: OBGRID config */
diff --git a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_subdev.h b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_subdev.h
index f388d3f..88d2e28 100644
--- a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_subdev.h
+++ b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_subdev.h
@@ -189,6 +189,17 @@ struct atomisp_css_params {
 	struct ia_css_shading_table *shading_table;
 	struct ia_css_morph_table   *morph_table;
 
+	/*Add new parameters for isp2.7*/
+	struct ia_css_dpc2_config dpc2_config;
+	struct ia_css_eed1_8_config eed1_8_config;
+	struct ia_css_ob2_config ob2_config;
+	struct ia_css_ctc2_config ctc2_config;
+	struct ia_css_iefd2_6_config iefd2_6_config;
+	struct ia_css_macc1_5_config macc1_5_config;
+	struct ia_css_macc1_5_table macc1_5_table;
+	struct ia_css_xnr3_0_11_config xnr3_0_11_config;
+
+
 	/*
 	 * Used to store the user pointer address of the frame. driver needs to
 	 * translate to ia_css_frame * and then set to CSS.
-- 
1.9.1

